charmcraft/templates/init/CONTRIBUTING.md.j2

```diff
@@ -0,0 +1,16 @@
+# {{ name }}
+
```

Comment:
May I make a couple of suggestions here keeping in view the following rationale -- In order to effectively contribute to a charm that has already reached a certain stage of maturity any developer will need to understand the following types of things

1. How is the code in this charm structured (a high level overview) ? For example the constructor of the charm makes clear what event are handled but it is not immediately obvious why only that particular subset of events were choosen to be handled and others ignored. This is but one example intended to highlight what is ment by code overview. Other obvious examples are -- What are the objects involved in this charm ? What are the responsibilities of these objects ? How do these objects collaborate ? Such understanding is essential for a new developer to understand the intent of the present design of the charm in order to be able to effectively extend it in the future.
2. What assumptions does it make on the use cases for this charm ? For example MongoDB charm only supports replicated databases. It does not support sharding for instance at least as yet. A developer or user should not have to read the entire charm code to understand what use cases are supported and what are not.
3. What were some key design decisions, specific to this charm ? There is a overlap here with point 2 above but this is focused on more the implementation than use cases. 

I think we will be missing a very important aspect of the documentation of charms if we do not encourage charm authors to document such "high level" things so that future contributors can easily understand this.

Whew! Hopefully I have made a case to request that the `CONTRIBUTING.md` file as a few additional headings such as

- Code overview
- Supported Use cases
- Design choices

etc.

Reply:
I'd also suggest including a Roadmap/TODO heading so people can identify if there are planned features they could contribute